### PR TITLE
Stabilise feature(never_type). Introduce feature(exhaustive_patterns)

### PR DESCRIPTION
**Pull request by canndrew**
This stabilizes `!`, removing the feature gate as well as the old defaulting-to-`()` behavior. The pattern exhaustiveness checks which were covered by `feature(never_type)` have been moved behind a new `feature(exhaustive_patterns)` gate.

### CODE REVIEW

File: src/libcore/cmp.rs [hunk 1]
```diff
@@ -880,24 +880,24 @@ mod impls {
 
     ord_impl! { char usize u8 u16 u32 u64 u128 isize i8 i16 i32 i64 i128 }
 
-    #[unstable(feature = "never_type", issue = "35121")]
+    #[stable(feature = "never_type", since = "1.24.0")]
```
Comment (varkor) [inline src/libcore/cmp.rs]: Shouldn't this be `since = "1.25.0"`?

Comment (canndrew) [reply src/libcore/cmp.rs]: True, it's too late to make 1.24. Thanks for the catch!

**Review by dtolnay — COMMENTED** [review]
I believe when we split a feature gate into stable part and unstable part, we typically try to rename the stable part rather than the unstable part. So in this case `!` would be stabilized under something like `feature(never_bang)` and what you have as `feature(exhaustive_patterns)` would continue to be `feature(never_type)`. This minimizes churn for libraries that currently use never_type. Their existing code would continue to compile.

**Comment by nikomatsakis** [comment]
If I understand correctly, this will also resolve the future compatibiltiy question (and make the change to fallback to `!` instead of `()`). Therefore, It seems like we should do a crater run to test the impact.

@bors try

**Comment by nikomatsakis** [comment]
If I understand correctly, this will also resolve the future compatibility question in https://github.com/rust-lang/rust/issues/39216 (and make the change to fallback to `!` instead of `()`). Therefore, It seems like we should do a crater run to test the impact.

@bors try

**Comment by SimonSapin** [comment]
> The pattern exhaustiveness checks which were covered by feature(never_type) have been moved behind a new feature(exhaustive_patterns) gate.

I haven’t followed all the discussions around `!`, could you give some examples of code affected by this? In this different from what’s already possible on stable with empty enums?

**Comment by SimonSapin** [comment]
Never mind, I saw your explenation on https://github.com/rust-lang/rust/issues/35121#issuecomment-360027186. As to empty enums, this also causes ``error[E0005]: refutable pattern in local binding: `Err(_)` not covered``

```rust
    enum Void {}
    let r: Result<u32, Void> = Ok(4);
    let Ok(i) = r;
```

**Comment by canndrew** [comment]
@dtolnay That would mean only the pattern-matching stuff is behind the `never_type` feature gate though, which would make it a bit of a misnomer.

As it is I deleted the `never_type` feature entirely. When I tried to add it back in order to follow https://forge.rust-lang.org/stabilization-guide.html it caused errors from anything tagged with `stable(feature = "never_type", since = "1.25.0")` complaining about mismatching issue numbers. There are no other `stable` tags with issue numbers so I don't know what it wants exactly.

**Comment by nikomatsakis** [comment]
@canndrew huh, I've never seen those particular errors before

**Comment by aidanhs** [comment]
Crater run started.

**Comment by aidanhs** [comment]
Crater run is currently stalled due to runs taking up more disk space than they used to. We've started up a bigger box and are working through the backlog - ETA another week or so, sorry :(

**Comment by canndrew** [comment]
@nikomatsakis 

> huh, I've never seen those particular errors before

Specifically, I added `(accepted, never_type, "1.25.0", Some(35121))` to the list of accepted features in features.rs. I then get tidy errors saying `mismatches to corresponding lang feature in: ["tracking issue"]` for all lines that say `#[stable(feature = "never_type", since = "1.25.0")]`

**Comment by aidanhs** [comment]
When this is ready for another run can you get bors to do a try? Then we'll prioritise this PR since it's been idle and waiting for a while.

**Comment by nikomatsakis** [comment]
@bors try

(I just rebased)

**Comment by nikomatsakis** [comment]
@aidanhs crater run requested

**Comment by aidanhs** [comment]
Sorry, was setting up a new crater box. Run started.

**Comment by canndrew** [comment]
So with 1.24 released is this going to make it into the beta for 1.25?

(I just noticed it needs a rebase, I'll get on that.)

**Comment by Mark-Simulacrum** [comment]
It is unlikely that we'll backport this stabilization; we generally avoid doing so.

**Comment by aidanhs** [comment]
Crater results: http://cargobomb-reports.s3.amazonaws.com/pr-47630/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/blacklist.md). If you see any spurious failures not on the list, please make a PR against that file.

**Comment by canndrew** [comment]
I've started looking into the crater regressions. A few of them are spurious (just tests randomly failing), and most the rest are due to this PR breaking clippy (due to `TyTuple` changing).

Of the remaining crates: some of them built/tested fine on my machine, even after running `cargo test` 10 times. I suspect these are just spurious failures as well, albeit very rares ones. A brief look at the crater results seems to back this up (because the failures involve scary-looking things like making network requests). These crates are `ntp-0.4.0`, `tasks-framework-0.1.0`, `sozu-lib-0.1.0`, `fix-rs-0.2.1`.

`panic-control-0.1.3` is failing in a doc test which is specifically testing for the new fallback behaviour (and expecting to not see it). It just needs to be patched.

`gdm_mi` I'm having trouble building on my machine. So I don't know what its story is.

And that just leaves the two actual regressions: `oplog-0.2.0` (actually it's its dependency `bson-0.3.2` which is broken) and `rspec-1.0.0-beta.4`. Both of these are relying on the old fallback behaviour, though luckily they're breaking at compile time, not run time. I'll submit PRs to fix these crates.

ping @nikomatsakis

**Comment by nikomatsakis** [comment]
@canndrew thanks for that analysis!

**Comment by pietroalbini** [comment]
@canndrew ping from the release team! What's the status on this PR?

**Comment by nikomatsakis** [comment]
@canndrew seems like we're going to land this soon, you up for a rebase? 

(What do you think about the improved error message, btw?)

**Comment by canndrew** [comment]
I've rebased it, though I'm [having trouble building from source](https://github.com/rust-lang/rust/issues/48612) at the moment so I'll have to wait for travis to say whether the rebase broke it or not.

> (What do you think about the improved error message, btw?)

Seems reasonable. So it would be a matter of changing `TyNever` to TyNever(bool)` then and using the flag to tell whether we should add a note to the error?

**Comment by canndrew** [comment]
Hopefully travis will pass this time. Should I add the error message change to this PR or allow this to be merged and add it in a follow-up PR? I should have time to implement it sometime in the next couple of days.

**Comment by nikomatsakis** [comment]
> So it would be a matter of changing TyNever to TyNever(bool) then and using the flag to tell whether we should add a note to the error?

Hmm, so, I'd prefer not to modify the `!` type (I want to keep the `sty` stuff clean). What I would prefer to do instead is this:

- When we report errors, pass in a boolean that indicates whether fallback *has* occurred
- If we have an error where the self type is `!`, and fallback has occurred, then:
    - check whether `()` would've worked using `evaluate_predicate`
    - report that 'in release 1.27.0, fallback changed from `()` to `!` for certain type variables; that may have caused this change'

or something like that?

I guess the question is whether that's too crude -- it seems unlikely to fire capriciously, but we shouldn't leave it in for *that* long probably.

**Comment by canndrew** [comment]
> When we report errors, pass in a boolean that indicates whether fallback has occurred

So, whenever we perform fallback we set a flag in a context somewhere saying that fallback has happened?

**Comment by nikomatsakis** [comment]
> So, whenever we perform fallback we set a flag in a context somewhere saying that fallback has happened?

Right, so what I meant is this. We do fallback right here:

https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/mod.rs#L876-L878

and immediately afterwards we select obligations here:

https://github.com/rust-lang/rust/blob/3eeb5a665e313c5b281820099e04d4c6c8188b46/src/librustc_typeck/check/mod.rs#L879

I was thinking that -- most likely -- errors that people are now getting because of fallback will occur at that point. i.e., this is where you *would* have had `()` but you now have `!`. So we could modify the method that reports errors around trait failures:

https://github.com/rust-lang/rust/blob/3eeb5a665e313c5b281820099e04d4c6c8188b46/src/librustc/traits/error_reporting.rs#L50-L52

to take an extra `bool` parameter indicating whether fallback has been performed. If so, it will emit the "better" error I described. We could then pass `true` for this parameter specifically for that one call to `select_obligations_where_possible` above, although I suppose it might be better  to have some flag in the `FnCtxt` that gets set to true when fallback has occurred, since we also select obligations at other points (though I doubt those would be affected, for the reason I just said: this first case is the one where I would expect fallback to have the maximal impact). 

I think that the code we need is probably very similar to the code we already have for future-compatibility, right? That is, where we currently check for whether `!` will *fail*, we would instead detect that `!` *has* failed and check whether `()` would've succeeded.

**Comment by mark-i-m** [comment]
@canndrew You appear to use the same commit message generation scheme I use!

**Comment by canndrew** [comment]
> I think that the code we need is probably very similar to the code we already have for future-compatibility, right? That is, where we currently check for whether `!` will fail, we would instead detect that `!` has failed and check whether `()` would've succeeded.

Right, I was just wondering how we get the information there about whether the `!` was formed by defaulting or not. I can't quite see the sequence of calls that go from `typeck_tables_of` to `SelectionContext::select` but I trust that there is one :). It seems like sticking it on the `TyNever` would be completely accurate though whereas anything else may not be, and it would only be a temporary thing. But I can thread a boolean argument through a sequence of calls instead if you'd prefer.

At any rate, should we merge this PR first since it's (finally) building? I will try to rush out a follow-up for the error message however we want to do it.

**Comment by canndrew** [comment]
How can I add a note to the error where it gets raised here?: https://github.com/rust-lang/rust/blob/0be38e1ce3dcfe6402b3bd9617f74857fe7fa8c3/src/librustc/traits/select.rs#L571

Before I could simply emit a warning there and then, but now since we're only adding a note when we have an error I need to somehow make that note appear on the error which I don't think gets emitted until later on. I could add a boolean flag on `SelectionError` that says "add a note to this error about the fallback change when emitting" but there must be an easier way.

**Comment by nikomatsakis** [comment]
@canndrew 

I think I expected to move that code into `traits::error_reporting`

**Comment by nikomatsakis** [comment]
@canndrew sorry if that was a bit brief -- did it make sense? Do you think you'll have time to get this patch done? I'd definitely like to land this stabilization =)

**Comment by clarcharr** [comment]
Would it be possible to include replacing `core::convert::Infalliable` to `!` in this PR, considering how this is likely to be merged before `try_from`?

If it makes more sense to include in a separate PR, I can provide one.

**Comment by canndrew** [comment]
@nikomatsakis I'd definitely like to land it too =). I've been very busy the last week however. I've added a commit for the error note, can you tell me if this is what you had in mind?

**Comment by canndrew** [comment]
@clarcharr I can throw in a commit for that when I have time.

**Review by nikomatsakis — COMMENTED** [review]
Yes, exactly what I had in mind, thanks!

File: src/librustc/lint/builtin.rs [hunk 2]
```diff
@@ -318,7 +304,6 @@ impl LintPass for HardwiredLints {
             DEPRECATED,
             UNUSED_UNSAFE,
             UNUSED_MUT,
-            COERCE_NEVER,
```
Comment (nikomatsakis) [inline src/librustc/lint/builtin.rs]: we also need to register that these lints are removed. Like so:

https://github.com/rust-lang/rust/blob/883e74645d350b6752cb94d48f46363f6f8789e9/src/librustc_lint/lib.rs#L316-L317

Presumably something like:

```rust
    store.register_removed("coerce_never",
        "converted into hard error, see https://github.com/rust-lang/rust/issues/XXX");
    store.register_removed("resolve_trait_on_defaulted_unit",
        "converted into hard error, see https://github.com/rust-lang/rust/issues/XXX");
```

File: src/librustc/middle/mem_categorization.rs [hunk 3]
```diff
@@ -1298,7 +1298,7 @@ impl<'a, 'gcx, 'tcx> MemCategorizationContext<'a, 'gcx, 'tcx> {
           PatKind::Tuple(ref subpats, ddpos) => {
             // (p1, ..., pN)
             let expected_len = match self.pat_ty(&pat)?.sty {
-                ty::TyTuple(ref tys, _) => tys.len(),
+                ty::TyTuple(ref tys) => tys.len(),
```
Comment (nikomatsakis) [inline src/librustc/middle/mem_categorization.rs]: so happy to see that `_` go =)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks great. Apart from the nits earlier, I had one concern, otherwise r=me.

**Comment by scottmcm** [comment]
@clarcharr My intuition is that `convert::Infallible` is part of `TryFrom` stabilization, not `!` stabilization, so a separate PR for it might be the way to go.

File: src/librustc/traits/error_reporting.rs [hunk 4]
```diff
@@ -534,9 +536,11 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
 
     pub fn report_selection_error(&self,
                                   obligation: &PredicateObligation<'tcx>,
-                                  error: &SelectionError<'tcx>)
+                                  error: &SelectionError<'tcx>,
+                                  fallback_has_occurred: bool)
     {
         let span = obligation.cause.span;
+        let _ = fallback_has_occurred;
```
[the remark this answers is not in the document]
Comment (canndrew) [reply src/librustc/traits/error_reporting.rs]: Oops, got left in from testing.

**Comment by canndrew** [comment]
@nikomatsakis I've addressed the nits. Is this ready to rock?

**Comment by nikomatsakis** [comment]
@bors r+

@canndrew 🎸 👏 🎸

**Comment by canndrew** [comment]
@clarcharr I've submitted a PR for the replacing `Infallible` with `!`: #49038

**Pull request merged**